daemon/indexing/apollo/index.js

```diff
@@ -4,37 +4,268 @@ var search = require('../lib/search.js')
 
 /*
- * Implentation of a GraphQL server using the Apollo framework.
- * See https://www.apollographql.com/server
+ * Implementation of the Origin GraphQL server.
+  * Uses the Apollo framework: https://www.apollographql.com/server
  *
  */
 
 // Type definitions define the "shape" of the data and specify
 // which ways the data can be fetched from the GraphQL server.
 const typeDefs = gql`
+  ######################
+  #
+  # Query output schema.
+  #
+  ######################
+
+  # When querying a set of items, the output is a page.
+  interface OutputPage {
+    totalCount: Int! # Total number of items in the set.
+    more: Boolean!   # True if there is another page to fetch.
+  }
+
+  type Price {
+    currency: String!
+    amount: String!
+  }
+
+  type User {
+    walletAddress: ID!   # Ethereum wallet address
+    identityAddress: ID  # ERC 725 identity address.
+    listings(page: Page, order: ListingOrder, filter: ListingFilter): ListingPage
+    offers(page: Page, order: OfferOrder, filter: OfferFilter): OfferPage
+    reviews(page: Page, order: ReviewOrder, filter: ReviewFilter): ReviewPage
+  }
+
+  enum OfferStatus {
+    CREATED
+    ACCEPTED
+    DISPUTED
+    # Note: There is no "Finalized" status stored on the Offer contract on-chain.
+    #       This is computed by the indexing server.
+    FINALIZED
+  }
+
+  type Offer {
+    id: ID!
+    ipfsHash: ID!
+    listingId: ID!
+    buyer: User!
+    status: OfferStatus!
+  }
+
+  type OfferPage implements OutputPage {
+    totalCount: Int!
+    more: Boolean!
+    offers: [Offer]
+  }
+
+  type Review {
+    ipfsHash: ID!
+    reviewer: User!
+    text: String!
+    rating: Int!
+  }
+
+  type ReviewPage implements OutputPage {
+    totalCount: Int!
+    more: Boolean!
+    reviews: [Review]
+  }
+
+  # TODO: Add a status indicating if Listing is sold out.
   type Listing {
-    id: String
-    name: String
+    id: ID!
+    ipfsHash: ID!
+    seller: User!
+    title: String!
     description: String
-    price: Float
+    category: String!
+    subCategory: String!
+    price: Price!
+    offers(page: Page, order: OfferOrder, filter: OfferFilter): OfferPage
+    reviews(page: Page, order: ReviewOrder, filter: ReviewFilter): ReviewPage
+  }
+
+  type ListingPage implements OutputPage {
+    totalCount: Int!
+    more: Boolean!
+    listings: [Listing]
+  }
+
+  ######################
+  #
+  # Query input schema.
+  #
+  # Note: Some input types have a "in" prefix because GraphQL does not allow to
+  # use thw same name for both an input and output type.
+  #
+  ######################
+  enum OrderDirection {
+    ASC   # Default if no direction specified.
+    DESC
+  }
+
+  input Page {
+    num: Int!  # Page number.
+    size: Int! # Number of items per page.
+  }
+
+  input inPrice {
+    currency: String!
+    amount: String!
+  }
+
+  #
+  # ORDER
+  #
+  enum OfferOrderField {
+    CREATION_DATE
+    STATUS
+  }
+
+  enum ReviewOrderField {
+    CREATION_DATE
```

Comment:
Maybe also sort by RATING_ASC and RATING_DESC

Reply:
Added.